### PR TITLE
[KEYCLOAK-14139] Upgrade login screen to PF4

### PR DESCRIPTION
**Pull request by mabartos**
JIRA: [KEYCLOAK-14139](https://issues.redhat.com/browse/KEYCLOAK-14139)

Design of new login screen with PF4 components. If there are less or equal than 4 identity providers, they are shown with full width as you can see below in mobile version. Title was changed to 'Log in to your account', because there would be 3 times title 'Log in' in a column. WDYT?

newLogin

Mobile version:
mobileNew

Custom icons for identity providers can be provided in `theme.properties`. 
You can see custom identity provider above with name `my-oidc`, so if you would like to add icon for that, just add line, like below, to `theme.properties` in login folder with pattern `kcLogoIdP-{alias}`. You can choose icons from PF4 icons website.

f.e `kcLogoIdP-my-oidc = fa fa-bell`

We are trying to complete task with PF4 upgrade ASAP (to the end of the current sprint). So, some design changes should be solved in a follow-up tasks.

Please @Pepo48, @vmuzikar , @ssilvert, could your review this? Thank you.

### CODE REVIEW

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties
Comment (ssilvert) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties:22]:
```suggestion
loginAccountTitle=Sign in to your account
```

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties
Comment (ssilvert) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties:23]:
```suggestion
loginTitle=Sign in to {0}
```

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties
Comment (ssilvert) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties:379]:
```suggestion
identity-provider-login-label=Or sign in with
```

**Review by ssilvert — CHANGES_REQUESTED** [review]
We have standardized on "Sign in" instead of "Log in".

I don't have time to do a thorough review but I can come back and do a better review later if you wish.

One thing you might want to take a closer look at is your use of CSS and hard-coded values.  If implemented properly, PatternFly can make it much easier for developers to customize look and feel.  I wouldn't hold up the PR because of this, but we will want to clean it up either now or in the future.

Here are the CSS guidelines we are using in the New Admin Console project: https://docs.google.com/document/d/1qVQ4KMt74CvsceV1cLr933cSj33oQIBICE7C6a3wnw8/edit#heading=h.b2okbjccloq7

**Comment by mabartos** [comment]
> We have standardized on "Sign in" instead of "Log in".
> 
Ok
> I don't have time to do a thorough review but I can come back and do a better review later if you wish.
> 
I understand. I think, it'll be sufficient, if @Pepo48 and @vmuzikar will review this.
> One thing you might want to take a closer look at is your use of CSS and hard-coded values. If implemented properly, PatternFly can make it much easier for developers to customize look and feel. I wouldn't hold up the PR because of this, but we will want to clean it up either now or in the future.
I'll try to take a look at that, but I'm not sure, how much time it'll take. As you said, it can be resolved in future, so some follow-up task can be created. Thanks @ssilvert

**Comment by vmuzikar** [comment]
@mabartos I've got a quick question before doing the review? Have you run the UI tests with your changes (both admin and account console)?

**Comment by mposolda** [comment]
@mabartos Thanks for the PR. Good work!

I did a quick check of sources and did not spot anything obvious. Few questions/concerns:
- Will be good to update documentation as well. I am adding label "missing-docs" due this. It seems that at least "Upgrading guide" worth mention that there is update to Patternfly 4. It worth to be added also to the "Release notes" IMO. And finally maybe some changes in the "Themes" chapter may be needed (not 100% sure, you can doublecheck if all the content would be still up-to-date once those changes are done)
- Will be good to update existing quickstarts and examples. I am not sure if we have any quickstarts for themes. However we have examples in the "examples/themes" and those worth updating. It may be also good exercise to update those, so you know what are the things required for the update
- All the admin/account UI tests will be good to doublecheck as @vmuzikar already pointed

Rather leaving to @Pepo48 and @vmuzikar for deeper review. I will approve once they approve and if my points above are addressed :-)

**Comment by vmuzikar** [comment]
I've triggered the build pipeline and post a link to test pipelines once they've started.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/LoginPasswordResetPage.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/LoginPasswordResetPage.java:79]: The test should be deterministic and should know whether it's fetching the error message from the notification or from the field.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/LoginPage.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/LoginPage.java:141]: The test should be deterministic and should know whether it's fetching the error message from the notification or from the field.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/PasswordPage.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/PasswordPage.java:62]: The test should be deterministic and should know whether it's fetching the error message from the notification or from the field.

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java
Comment (vmuzikar) [inline services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java:130]: Should this really be a field error? I mean a disabled account is not an input error, so why it should be assigned to a field?

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/PasswordPage.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/PasswordPage.java:54]: The WebElements injected by Selenium will never be null.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Thank you @mabartos, good work! :) It's a nice upgrade.

Added some comments, mainly only nitpicks.

Also, should we have some migration steps in case someone uses the old login theme as a base for their theme? @mposolda?

I've got some good news. :D The UI tests seem to be ok (no new failures caused by this): https://keycloak-jenkins.com/job/universal-test-pipeline-server/2186/

File: themes/src/main/resources/theme/base/login/login.ftl
Comment (mabartos) [reply themes/src/main/resources/theme/base/login/login.ftl:1]: Yes, you're right. The input box is different, and I didn't even notice that. The background overflow is common issue in KC. For the future, it needs some polishing.Thanks @vmuzikar !

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: It's not intended. It needs some predicate, which helps to avoid the empty info message. Thanks!

File: themes/src/main/resources/theme/base/login/login.ftl
Comment (mabartos) [reply themes/src/main/resources/theme/base/login/login.ftl:2]: I've thought, that was intended for tablet resolution. But, if I take a look at this now, we don't need this. We've dropped the side IdPs, so it's not necessary anymore.

**Review by vmuzikar — COMMENTED** [review]
@mabartos I've noticed one more possible issue. In your screenshot the IdPs are nicely placed in two columns whereas for me it is always only one column:
Screenshot 2020-10-13 at 09 27 19
Is this intended? IMHO two columns look better. ;)

**Comment by mposolda** [comment]
> Also, should we have some migration steps in case someone uses the old login theme as a base for their theme? @mposolda?

Yes, I believe we need something in the migration guide as pointed in my comment above. I am not exactly sure what it should be and what would be the typical changes, which people will need to do when they migrate their login theme.

Because of that, I proposed to do this exercise in our theme quickstarts (if we have any) and in the theme examples in the "themes/examples" . If those require some updates, it is high chance that custom themes by customer will require them as well.

IMO in minimum, we should add to the migration guide just something like "We updated login themes to PAtternfly 4. Please upgrade your custom themes accordingly" without any other info. But if we can add some additional info/hints, it might be beneficial.

In case that migration would be an effort for people, it can be ideal if there is a way for customer to eventually use the "old" Patternfly, and just do something easy in his themes to fallback to the old style if needed. However I am not sure if this is easily possible?

**Comment by mabartos** [comment]
> @mabartos I've noticed one more possible issue. In your screenshot the IdPs are nicely placed in two columns whereas for me it is always only one column:
> Is this intended? IMHO two columns look better. ;)

Yes, this was intended. I kept the count of IdPs, when the layout is changed. If you have more than 4 IdPs, you should see 2 columns. I've just tried to decrease the number to 3 and it looks maybe better. I you have 4 IdPs, you'll see 2 columns and each column contains 2 items. I'll keep the version with 3 IdPs. Thanks @vmuzikar !

**Comment by vmuzikar** [comment]
@mabartos I had 4 IdPs (check my screenshot) and only 1 column. ;)

**Comment by mabartos** [comment]
> @mabartos I had 4 IdPs (check my screenshot) and only 1 column. ;)

@vmuzikar I'm sorry for misunderstanding. Currently, you have to provided more than 4 IdPs, in order to see the 2 columns. Previously, I meant the case, where the count of IdPs, which change layout, is 3, which will be in more recent version.

**Comment by mabartos** [comment]
I've just created a PR to documentation (See [#1050](https://github.com/keycloak/keycloak-documentation/pull/1050)). All request changes should be resolved. I've also created a follow-up task for quickstarts/examples (See [KEYCLOAK-15962](https://issues.redhat.com/browse/KEYCLOAK-15962)). 

Please, could you restart Travis? There was some error with too many requests on the github server. Thanks!

**Comment by mabartos** [comment]
> Yes, I believe we need something in the migration guide as pointed in my comment above. I am not exactly sure what it should be and what would be the typical changes, which people will need to do when they migrate their login theme.
> 
> Because of that, I proposed to do this exercise in our theme quickstarts (if we have any) and in the theme examples in the "examples/themes" . If those require some updates, it is high chance that custom themes by customer will require them as well.
> 
@mposolda Thanks for your comment! I have prepared PR, which updates theme example, where the required changes are shown. But, these are rather minor changes, which shouldn't affect any custom themes so much. So, I think, it's not a big deal for customers.

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (pdrozd) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar +1, this should work as well

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @pdrozd Just to be clear. You mean IE should work or the idea of dropping IE should work? :)

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar @pdrozd Thanks for your comments. 

IE doesn't support `var()` function, so it's not possible to load the variables. In this case, there should be some default hardcoded values in main `.css` file. But, it's contradicted against what @ssilvert suggested above. Then, there will be duplicate properties in  the main `.css` file, in order to keep compatibility with various browsers. However, it's true, the login screen should be displayed properly on all browsers, so I'll add the default values. I just wanted to highlight this issue.

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: Just for the record, here's an [official guide for IE11 support in PF4](https://github.com/patternfly/patternfly/wiki/IE11-Support).

Comment (pdrozd) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar I meant that It would be good if this page also work with IE. The support for IE is dropped, so if this is really complicated, I would not insist on this.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I've created some workaround for that and it should be working with IE. But there are another screens, which should be remade then and I think, it's not worth the effort.

**Comment by mposolda** [comment]
@mabartos Thanks! I will approve if @vmuzikar is fine with this PR :-)

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (ssilvert) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: > @vmuzikar I meant that It would be good if this page also work with IE. The support for IE is dropped, so if this is really complicated, I would not insist on this.

@pdrozd @vmuzikar @mabartos Can we say for sure that IE support is dropped for login screens?  That sounds like a wider discussion.

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @ssilvert I think we cannot. AFAIK we only dropped IE support in Admin and new Account Consoles. If we want to drop the support for login pages as well, I think it should be blessed by PM and maybe discussed on dev ML as well.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I'd say, there could be some basic (not like with PF4) functionality/appearance in login screens and inform customer/user, that the IE support will be dropped out soon and that there is a need for using other browser. I mean, there could be some warning alert with the message. At this time, IMHO, it's not a big deal for customer/user switch to another browser. AFAIK, IE is barely supported by Microsoft, so I think, there is no need for us support that in the big manner either. WDYT?

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @mabartos +1 but I really think that's not your call and we need to involve PM into the discussion.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar I agree. So, what do you suggest, when we'd like to include this in KC 12? Keep some default values, in order to provide, at least, minimal support for IE and then, after discussion, remove/keep the default values? Or is it blocker for this?

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I see it as a blocker. We still need to count with the possibility that it needs to work in IE. And currently, some pages are broken in IE as you mentioned above, am I correct?

Comment (mposolda) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I vote for add any hacks needed for the IE support. And create follow-up JIRA "Remove hacks after IE support is removed" for the future :-)

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar @mposolda Thanks guys! This login screen should work properly with IE, but another screens, which will be edited lately, are probably broken and we'll add the "hack" for solving that. :)

**Comment by Pepo48** [comment]
@mabartos (Cc: @mposolda ) the proposed changes LGTM as well, thanks everyone for suggestions.
I need to check the IE behaviour with the workaround one more time, but overall very good job!

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @mabartos Do you plan to address the IE issue in this PR or a follow-up?

Comment (mposolda) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar @mabartos Correct me if I am wrong, but ideally it will be good to include everything in this PR? My point is, that if IE fix is done as a follow-up, then there will be some time period during which the Keycloak master will be broken with IE. Hence there is a risk of scenario like:
1) This PR is merged . KEycloak master will become broken with IE
2) Follow-up PR is send for fix IE
3) Keycloak 12 is released
4) PR from step 2 is merged. But it is late as Keycloak 12 was already released with broken IE...

(same can happen with steps 2 or 3 switched)

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @mposolda +1

Comment (Pepo48) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: > The page is completely broken in IE. We're starting to drop IE support, but not sure if we're ready to do so on something as elementary as the login screen. 

@mabartos (Cc: @vmuzikar @mposolda) after adding the workaround the screen looks good even on IE11:
image

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: If this PR works correctly with IE, there won't be error with another screens. PF3 and PF4 work simultaneously, so some of the additional PRs don't have to be merged, if they have some issue with IE yet. The additional PRs only upgrade the components, which look, in most cases, very similar. This PR will work with IE, and I think, it's sufficient so far. So there won't be any follow-up task for IE workaround. Everything will be included in each PR.

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @Pepo48 @mabartos From my PoV it feels kinda broken in IE. Just compare the following screenshots (left is PF3 in IE, right PF4 in IE):
image
image-2
Notice the input fields which I think are missing some padding, notification looks akward. And it's possible there'll be some more broken components. I think it's definitely a downgrade in UX for IE users.

But I don't want to block this so I'll leave the decision to others if this is good enough or not.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I'll try to remake it, but I can't spend a lot of time on that.

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @mabartos Thanks!

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: It isn't like on Chrome or other browsers, but I think, it's sufficient. WDYT @vmuzikar ?
ieSignIn

**Comment by mabartos** [comment]
I've created the workaround for IE. Unfortunately, Travis has failed after 40 seconds with response code `529 Too many requests`, but it wasn't caused by this PR approach.

File: themes/src/main/resources/theme/base/login/template.ftl
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @mabartos Thank you for another batch of fixes! Nice job! I know IE can be very frustrating sometimes...

The only issue I see is the notifications and how the icon is somewhat misaligned:
Capture

But again, I don't want to block this PR because of this. I'll leave it to other if we can live with this minor UI glitch.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: @vmuzikar Thanks for your review. I know, there is some issue with register alert (some alert with longer message). But in the follow-up task with registration, there won't be this alert, but the errors will be under the inputs. Otherwise, the simple alert looks almost normal. What about edit these minor changes in follow-up task, which I'm currently working on in order to not block other PRs, which depend on this PR?

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I think the misaligned notification icons affect all notifications, not just alerts with a longer message. It's apparent on your screenshot above too.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: Yes, you're right. All notifications.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/template.ftl:2]: I was able to edit the notifications. At this moment, I think, the IE compatibility is solved for this PR. So, IMO, it could be finally merged. WDYT @vmuzikar ? :) 

ieSignInSuccess
ieREgisterErr

**Review by vmuzikar — APPROVED** [review]
@mabartos Thank you, looks good now!

Found one last tiny bug. The IdP are not placed into two columns in IE:
Capture
I see it as a non-blocker, so I'm going to approve now (I'll be OOO till monday, so you don't need to wait for me). Fix it if you find the time to do it.

**Comment by mabartos** [comment]
> @mabartos Thank you, looks good now!
> I see it as a non-blocker, so I'm going to approve now (I'll be OOO till monday, so you don't need to wait for me). Fix it if you find the time to do it.

@vmuzikar Thanks for approving! I was looking at it and it's not so easy to do that on IE for me. It's only minor thing and I think, it's not worth the big effort :)

**Review by mposolda — APPROVED** [review]
@mabartos @vmuzikar @Pepo48 @ssilvert Thanks everyone for the hard work on adding and reviewing this!

**Comment by Pepo48** [comment]
@mabartos (cc: @mposolda @vmuzikar @ssilvert) just a small "after-merge tip" that might be considered as a follow-up task - IE11 CSS workarounds can be targeted purely just for IE11 browser. I did basically the same thing here: https://github.com/keycloak/keycloak/pull/7516/files#diff-38231fba93cd3f639b39102b61bfa34322fa029a0f860611a537db1cc4101de2R3

In your case it probably requires small structural changes of the css though.

**Comment by mabartos** [comment]
@mposolda @Pepo48 @vmuzikar @ssilvert @pdrozd Thank you from me as well. I appreciate that very much :))

**Comment by mposolda** [comment]
> just a small "after-merge tip" that might be considered as a follow-up task - IE11 CSS workarounds can be targeted purely just for IE11 browser. I did basically the same thing here: https://github.com/keycloak/keycloak/pull/7516/files#diff-38231fba93cd3f639b39102b61bfa34322fa029a0f860611a537db1cc4101de2R3

@Pepo48 @mabartos Does this worth to create a follow-up JIRA for this task? If so, can you please create it?

**Pull request merged**
